File: examples/javaee/src/main/java/grooves/example/javaee/queries/PatientHealthQuery.java
package grooves.example.javaee.queries;

import com.github.rahulsom.grooves.api.EventApplyOutcome;
import com.github.rahulsom.grooves.java.Query;
import grooves.example.javaee.Database;
import grooves.example.javaee.domain.Patient;
import grooves.example.javaee.domain.PatientCreated;
import grooves.example.javaee.domain.PatientHealth;
import grooves.example.javaee.domain.ProcedurePerformed;
import lombok.Getter;
import org.jetbrains.annotations.NotNull;
import org.reactivestreams.Publisher;

import javax.inject.Inject;

import static com.github.rahulsom.grooves.api.EventApplyOutcome.*;
import static rx.Observable.*;
import static rx.RxReactiveStreams.*;

@Query(aggregate = Patient.class, snapshot = PatientHealth.class)
public class PatientHealthQuery implements CustomQuerySupport<PatientHealth> {

    @Inject @Getter
    private Database database;

    @Override
    public Class<PatientHealth> getSnapshotClass() {
        return PatientHealth.class;
    }

    @NotNull
    @Override
    public PatientHealth createEmptySnapshot() {
        return new PatientHealth();
    }

    @Override
    public void addToDeprecates(
            @NotNull PatientHealth snapshot, @NotNull Patient deprecatedAggregate) {
        snapshot.getDeprecates().add(deprecatedAggregate);
    }

    @Override
    public Publisher<EventApplyOutcome> applyPatientCreated(
            PatientCreated event, PatientHealth snapshot) { // <12>
        if (snapshot.getAggregate() == event.getAggregate()) {
            snapshot.setName(event.getName());
        }
        return toPublisher(just(CONTINUE)); // <13>
    }


    @Override
    public Publisher<EventApplyOutcome> applyProcedurePerformed(
            ProcedurePerformed event, PatientHealth snapshot) {
        snapshot.getProcedures().add(
                new PatientHealth.Procedure(event.getCode(), event.getTimestamp()));
        return toPublisher(just(CONTINUE));
    }

}

File: examples/springboot/kotlin/src/main/kotlin/grooves/boot/kotlin/queries/PatientHealthQuery.kt
package grooves.boot.kotlin.queries

import com.github.rahulsom.grooves.api.EventApplyOutcome.CONTINUE
import com.github.rahulsom.grooves.queries.QuerySupport
import com.github.rahulsom.grooves.queries.internal.SimpleExecutor
import com.github.rahulsom.grooves.queries.internal.SimpleQuery
import grooves.boot.kotlin.domain.Patient
import grooves.boot.kotlin.domain.PatientEvent
import grooves.boot.kotlin.domain.PatientHealth
import grooves.boot.kotlin.domain.Procedure
import grooves.boot.kotlin.repositories.PatientEventRepository
import grooves.boot.kotlin.repositories.PatientHealthRepository
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Component
import reactor.core.publisher.Mono.just
import java.lang.Exception
import java.util.*

@Component
class PatientHealthQuery :
        QuerySupport<Patient, String, PatientEvent, String, PatientHealth>,
        SimpleQuery<Patient, String, PatientEvent, PatientEvent.Applicable, String, PatientHealth> {

    override fun getExecutor() = SimpleExecutor<Patient, String, PatientEvent,
            PatientEvent.Applicable, String, PatientHealth, PatientHealthQuery>()

    @Autowired lateinit var patientEventRepository: PatientEventRepository
    @Autowired lateinit var patientHealthRepository: PatientHealthRepository

    override fun createEmptySnapshot() = PatientHealth()

    override fun getSnapshot(maxPosition: Long, aggregate: Patient) =
            patientHealthRepository.findByAggregateIdAndLastEventPositionLessThan(
                    aggregate.id!!, maxPosition)

    override fun getSnapshot(maxTimestamp: Date, aggregate: Patient) =
            patientHealthRepository.findByAggregateIdAndLastEventTimestampLessThan(
                    aggregate.id!!, maxTimestamp)

    override fun shouldEventsBeApplied(snapshot: PatientHealth) = true

    override fun addToDeprecates(snapshot: PatientHealth, deprecatedAggregate: Patient) {
        snapshot.deprecatesIds.add(deprecatedAggregate.id!!)
    }

    override fun onException(e: Exception, snapshot: PatientHealth, event: PatientEvent) =
            just(CONTINUE)

    override fun getUncomputedEvents(
            aggregate: Patient, lastSnapshot: PatientHealth?, version: Long) =
            patientEventRepository.
                    findAllByPositionRange(
                            aggregate.id!!, lastSnapshot?.lastEventPosition ?: 0, version)

    override fun getUncomputedEvents(
            aggregate: Patient, lastSnapshot: PatientHealth, snapshotTime: Date) =
            lastSnapshot.lastEventTimestamp?.
                    let {
                        patientEventRepository.
                                findAllByTimestampRange(
                                        aggregate.id!!, it, snapshotTime)
                    } ?:
                    patientEventRepository.
                            findAllByAggregateIdAndTimestampLessThan(aggregate.id!!, snapshotTime)

    override fun applyEvent(event: PatientEvent.Applicable, snapshot: PatientHealth) =
            when (event) {
                is PatientEvent.Applicable.Created -> {
                    if (event.aggregateId == snapshot.aggregateId) {
                        snapshot.name = event.name
                    }
                    just(CONTINUE)
                }
                is PatientEvent.Applicable.ProcedurePerformed -> {
                    snapshot.procedures.add(Procedure(event.code, event.timestamp))
                    just(CONTINUE)
                }
                is PatientEvent.Applicable.PaymentMade -> {
                    just(CONTINUE)
                }
            }
}